src/main/java/frc/robot/Constants.java

```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
+        //gains
+        public static final Gains LEFT_GAINS = new Gains(0, 0, 0);
+        public static final Gains RIGHT_GAINS = new Gains(0, 0, 0);
+
+        //movement limitation
+        public static final double kMaxLinearVelocity = 0;
+        public static final double kMaxLinearAcceleration = 0;
+
+        public static final double kGearing = 0;
+        public static final double kDriveWheelRaius = 0;
+
+        public static final double kRotToMeter = 2 * PI * kDriveWheelRaius * kGearing;
+        public static final double kRotToMeterPerSec = kRotToMeter / 60;
+        public static final int leftID = 0;
+        public static final int rightID = 0;
+        public static final double LEFT_ARM_LOCATION = 0;
+        public static final double RIGHT_ARM_LOCATION = 0;
+        public static final double MINIMAL_HEIGHT = 0;
+        public static final double EXTRA_SAFETY_DISTANCE = 0.1;
+
+        public enum Chain {
+            CHAIN_0(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_120(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_240(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_60(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_180(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0)),
+            CHAIN_300(new Translation2d(0, 0), new Translation2d(0, 0), new Translation2d(0, 0));
+
+            private final Translation2d negEdge;
+            private final Translation2d posEdge;
+            private final Translation2d centerStage;
+
+            Chain(Translation2d negEdge, Translation2d posEdge, Translation2d centerStage) {
+                this.negEdge = negEdge;
+                this.posEdge = posEdge;
+                this.centerStage = centerStage;
+            }
+
+            public Translation2d getNegEdge() {
+                return negEdge;
+            }
+
+            public Translation2d getPosEdge() {
+                return posEdge;
+            }
+
+            private boolean inRange(Translation2d robotTranslation) {
+                double robotStageAngle =
+                        Math.atan(robotTranslation.minus(centerStage).getX() / robotTranslation.minus(centerStage).getY());
+                double negAngle = Math.atan(negEdge.minus(centerStage).getX() / negEdge.minus(centerStage).getY());
+                double posAngle = Math.atan(posEdge.minus(centerStage).getX() / posEdge.minus(centerStage).getY());
+
+                if (posAngle > negAngle) return robotStageAngle < posAngle && robotStageAngle > negAngle;
+                return (robotStageAngle > posAngle && robotStageAngle > negAngle)
+                        || (robotStageAngle < posAngle && robotStageAngle < negAngle);
+            }
+
+            public static Chain getBestChain(Translation2d robotTranslation) {
```

Comment:
1. Documentation documentation documentation.
2. Early returns might be preferable here than a result variable, but that might be personal preference/style. I prefer it because I find it easier to track when nothing can change the result after it's set.
3. What happens if this function is called from the other side of the field? Is that what we want to happen?